File: find_pair/src/python/find_pairs.py
# This fails when we have too many recursive calls
def find_pairs_recursively(array, target_sum):
    def check_next_elem(elem, rest):
        if len(rest) < 1:
            print("\tGot to end of array.")
            return
        need = target_sum - elem
        if need in rest:
            print("\tPair: {} and {}".format(elem, need))
        else:
            check_next_elem(rest[0], rest[1:])

    check_next_elem(array[0], array[1:])


def find_pairs_loop(array, target_sum):
    for (i, x) in enumerate(array):
        if (target_sum-x) in array[i:]:
            print("\tPair: {} and {}".format(x, target_sum-x))
            break


# Driver to test
if __name__ == "__main__":
    arr = range(5, 500000000, 3)
    target_sum = 3698923
    import timeit

    print("Loop:")
    loopTime = timeit.timeit(lambda: find_pairs_loop(arr, target_sum), number=1)
    print("\tExecution Time: {0:.5f}".format(loopTime))
    # Loop:
    # Pair: 5 and 3698918
    # Execution Time: 0.00001

    print()
    print("Recursive:")
    recursiveTime = timeit.timeit(lambda: find_pairs_recursively(arr, target_sum), number=1)
    print("\tExecution Time: {0:.5f}".format(recursiveTime))
    # Recursive:
    # Pair: 5 and 3698918
    # Execution Time: 0.00001

File: find_pair/src/scala/FindPairs.scala

object FindPairs extends App {
  def findPairsRecursively(seq: Seq[Int], target_sum: Int): String = {
    @scala.annotation.tailrec
    def checkNextElem(elem: Int, rest: Seq[Int]): String = {
      rest.isEmpty match {
        case true => "No Pairs found"
        case _ =>
          val need = target_sum - elem
          if (rest.contains(need)) s"Pair: $elem and $need"
          else checkNextElem(rest.head, rest.tail)
      }
    }
    checkNextElem(seq.head, seq.tail)
  }

  def findPairsInLoop(seq: Seq[Int], target_sum: Int): String = {
    for (elem <- seq) {
      if (seq.tail.contains(target_sum-elem))
        return s"Pair: $elem and ${target_sum-elem}"
    }
    "No Pairs found"
  }

  def measureCall(f: (Seq[Int],Int) => String, s: Seq[Int], i: Int): Unit = {
    val startTime = System.nanoTime
    println(s"\t${f(s,i)}")
    val micros = (System.nanoTime - startTime) / 1000
    println(f"\tExecution Took: ${micros}%,d MicroSeconds")
  }

  val seq = 1 to 50000000 by 5
  val target_sum = 347332
  println("Recursive:")
  measureCall(findPairsRecursively, seq, target_sum)
//  Recursive:
//    Pair: 1 and 347331
//  Execution Took: 868 MicroSeconds
  println("Loop:")
  measureCall(findPairsInLoop, seq, target_sum)
//  Loop:
//    Pair: 1 and 347331
//  Execution Took: 2,336 MicroSeconds
}
